tracer startup logs in troubleshooting page (#7901)
DataDog/documentation: content/en/tracing/troubleshooting/tracer_startup_logs.md

---
title: Tracer Startup Logs
kind: Documentation
---
## Startup Logs

Tracer startup logs will capture all obtainable information at startup and log it either as `DATADOG TRACER CONFIGURATION` or `DATADOG TRACER DIAGNOSTICS` for easy searching within your logs.  Note some languages may log to a separate file depending on language conventions and the safety of accessing `Stdout` or equivalent.  In those cases, the location of logs will be noted in the respective language tab below.

> `DIAGNOSTICS` log lines appear when the tracer encountered an error during application startup, while  `CONFIGURATION` logs are a JSON or raw-formatted representation of settings applied to your tracer. In languages where an Agent connectivity check can be performed, the configuration JSON will also include a key 'agent_error', which can also be checked to see if the Agent is reachable.

Reviewer: ```suggestion
`DIAGNOSTICS` log entries happen when the tracer encounters an error during application startup, while  `CONFIGURATION` logs are a JSON or raw-formatted representation of settings applied to your tracer. In languages where an Agent connectivity check is performed, the configuration JSON will also include an 'agent_error' key, which indicates whether the Agent is reachable.
```

> If you see any `DIAGNOSTICS` log lines, please confirm from the indicated log that settings and configurations have been applied correctly.  If you do not see logs at all, please ensure that your application logs are not silenced and that your log level is at least `INFO` where applicable.

Reviewer: ```suggestion
If you see any `DIAGNOSTICS` log lines, confirm from the indicated log that settings and configurations are applied correctly.  If you do not see logs at all, ensure that your application logs are not silenced and that your log level is at least `INFO` where applicable.
```

{{< tabs >}}
{{% tab "Java" %}}

#### Configuration

```text
{"os_name":"Mac OS X","os_version":"10.15.4","architecture":"x86_64","lang":"jvm","lang_version":"11.0.6","jvm_vendor":"AdoptOpenJDK","jvm_version":"11.0.6+10","java_class_version":"55.0","enabled":true,"service":"unnamed-java-app","agent_url":"http://localhost:8126","agent_error":false,"debug":false,"analytics_enabled":false,"sampling_rules":[{},{}],"priority_sampling_enabled":true,"logs_correlation_enabled":false,"profiling_enabled":false,"dd_version":"null","health_checks_enabled":false,"configuration_file":"no config file present","runtime_id":"b69deb26-8bc3-4c00-8952-d42bf8c2123b"}
```

#### Diagnostics

> The Java tracer does not currently output any Diagnostics logs.  For this check, please run the tracer in [debug mode][1].

Reviewer: ```suggestion
The Java tracer does not output Diagnostics logs.  For this check, run the tracer in [debug mode][1].
```

[1]: /tracing/troubleshooting/tracer_debug_logs/
{{% /tab %}}
{{% tab ".NET" %}}

#### File Location

> Log files will be located within `%PROGRAMDATA%\Datadog .NET Tracer\logs`.

Reviewer: ```suggestion
Log files are located in `%PROGRAMDATA%\Datadog .NET Tracer\logs`.
```

- `dotnet-tracer-{processName}-{timestamp}.log` contains the Configuration log.

- `dotnet-profiler.log` contains the Diagnostics logs, if any are generated.

#### Configuration

```text
2020-06-29 12:26:39.572 +02:00 [INF] DATADOG TRACER CONFIGURATION -
{"date":"2020-06-29T12:26:39.5615724+02:00","os_name":"Windows",
"os_version":"Microsoft Windows NT 6.2.9200.0","version":"1.17.1.0",
"platform":"x64","lang":".NET Framework","lang_version":"4.8+",
"env":null,"enabled":true,"service":"git-credential-manager",
"agent_url":"http://localhost:8126","debug":false,
"analytics_enabled":false,"sample_rate":null,"sampling_rules":null,
"tags":[],"log_injection_enabled":false,
"runtime_metrics_enabled":false,"disabled_integrations":[]}
```

#### Diagnostics

> The .NET tracer has many diagnostic lines it can print.  Please see below for a complete list:

Reviewer: ```suggestion
The .NET tracer prints the following diagnostic lines:
```

```text
[2020-07-01T17:42:50Z] DATADOG TRACER CONFIGURATION - {"agent_error":"Couldn't connect to server","ddtrace.request_init_hook_reachable":false,"date":"2020-07-01T17:42:50Z","os_name":"Linux 49b1cb4bdd12 4.19.76-linuxkit #1 SMP Tue May 26 11:42:35 UTC 2020 x86_64","os_version":"4.19.76-linuxkit","version":"1.0.0-nightly","lang":"php","lang_version":"7.4.5","env":null,"enabled":true,"service":null,"enabled_cli":false,"agent_url":"https://localhost:8126","debug":false,"analytics_enabled":false,"sample_rate":1.000000,"sampling_rules":null,"tags":null,"service_mapping":null,"distributed_tracing_enabled":true,"priority_sampling_enabled":true,"dd_version":null,"architecture":"x86_64","sapi":"cgi-fcgi","ddtrace.request_init_hook":null,"open_basedir_configured":false,"uri_fragment_regex":null,"uri_mapping_incoming":null,"uri_mapping_outgoing":null,"auto_flush_enabled":false,"generate_root_span":true,"http_client_split_by_domain":false,"measure_compile_time":true,"report_hostname_on_root_span":false,"traced_internal_functions":null,"auto_prepend_file_configured":false,"integrations_disabled":null,"enabled_from_env":true,"opcache.file_cache":null}
```

#### Diagnostics

> The Diagnostics for the PHP tracer will print if the tracer is in DEBUG mode. They will be available through the other options below.

Reviewer: ```suggestion
The Diagnostics for the PHP tracer print if the tracer is in DEBUG mode. They are available through the other options below.
```

Reviewer: but also, it's unclear to me what "available through the other options below" means

Author: PHP Info section, or CLI SAPI, etc. all the remaining sections are just ways to get the same information

```text
[2020-07-01T17:35:25Z] DATADOG TRACER DIAGNOSTICS - agent_error: Couldn't connect to server
[2020-07-01T17:35:25Z] DATADOG TRACER DIAGNOSTICS - ddtrace.request_init_hook_reachable: false
```

#### PHP Info

> The JSON string can also be obtained from the PHP info page and the diagnostic information will be displayed in a separate table to help diagnose common issues.

Reviewer: ```suggestion
Obtain the JSON string from the PHP info page. Diagnostic information is be displayed in a separate table to help diagnose common issues.
```

Reviewer: This information doesn't seem clear to me, but I'm not sure what it should say.

{{< img src="tracing/troubleshooting/PHPInfo.png" alt="PHP Info"  >}}

#### CLI SAPI

> From the CLI SAPI, the info can be obtained from `php --ri=ddtrace`.

Reviewer: ```suggestion
Get the info from the CLI SAPI by running `php --ri=ddtrace`.
```

                               Diagnostics
agent_error => Couldn't connect to server
ddtrace.request_init_hook_reachable => false

Directive => Local Value => Master Value
ddtrace.disable => Off => Off
...
```

#### Runtime

> The JSON string can also be accessed at runtime from `\DDTrace\startup_logs()`.

Reviewer: ```suggestion
Access the JSON string at runtime with `\DDTrace\startup_logs()`.
```

```php
echo \DDTrace\startup_logs() . PHP_EOL;
```
{{% /tab %}}

{{% tab "Go" %}}

#### Configuration

```text
2020/07/09 15:57:07 Datadog Tracer v1.26.0 INFO: DATADOG TRACER CONFIGURATION {"date":"2020-07-09T15:57:07-05:00","os_name":"darwin","os_version":"10.15.4","version":"v1.26.0","lang":"Go","lang_version":"go1.14.2","env":"","service":"splittest2","agent_url":"http://127.0.0.1:8126/v0.4/traces","agent_error":"","debug":true,"analytics_enabled":false,"sample_rate":"NaN","sampling_rules":null,"sampling_rules_error":"","tags":{"runtime-id":"d269781c-b1bf-4d7b-9a55-a8174930554f"},"runtime_metrics_enabled":false,"health_metrics_enabled":false,"dd_version":"","architecture":"amd64","global_service":""}
```

#### Diagnostics

> The NodeJS Tracer has two possible diagnostic lines it may print, one in the case the agent cannot be reached, and the other in the case of trace sampling errors.

Reviewer: ```suggestion
The NodeJS Tracer prints one of two possible diagnostic lines, one for when the Agent cannot be reached, and the other for trace sampling errors.
```

{{% tab "NodeJS" %}}

#### Configuration

```text
[2020-07-02 14:51:16.421] [INFO] app - host:port==localhost:9080
[2020-07-02 14:51:16.423] [INFO] app - db type==mongo
[2020-07-02 14:51:16.439] [INFO] routes - Use dataaccess:../dataaccess/mongo/index.js
[2020-07-02 14:51:16.599] [INFO] app - Error connecting to database - exiting process: MongoError: connect ECONNREFUSED 127.0.0.1:27017
[2020-07-02 14:51:16.600] [INFO] app - Initialized database connections
[2020-07-02 14:51:16.601] [INFO] app - Express server listening on port 9080
DATADOG TRACER CONFIGURATION - {"date":"2020-07-02T18:51:18.294Z","os_name":"Darwin","os_version":"19.2.0","architecture":"x64","version":"0.22.0","lang":"nodejs","lang_version":"12.18.1","enabled":true,"service":"acmeair","agent_url":"http://localhost:8126","agent_error":"Network error trying to reach the agent: connect ECONNREFUSED 127.0.0.1:8126","debug":false,"analytics_enabled":false,"sample_rate":1,"sampling_rules":[],"tags":{"service":"acmeair","version":"0.0.4"},"dd_version":"0.0.4","log_injection_enabled":false,"runtime_metrics_enabled":false,"integrations_loaded":["http","fs","net","dns","express@4.17.1"]}
```

#### Diagnostics

> The NodeJS Tracer has one diagnostic line it may print, in the case the agent cannot be reached.

Reviewer: ```suggestion
The NodeJS Tracer prints a diagnostic line when the Agent cannot be reached.
```

```text
DATADOG TRACER DIAGNOSTIC - Agent Error: Network error trying to reach the agent: connect ECONNREFUSED 127.0.0.1:8126
```

{{% /tab %}}
{{% tab "Python" %}}

#### Log location

> If an existing logging configuration is found, the Python tracer will logs the configuration as INFO-level.  If there are any diagnostics those are logged as ERROR.

Reviewer: ```suggestion
The Python tracer logs configuration information, if found, as INFO-level. It logs diagnostics as ERROR.
```

> If there is no logging configuration, only Diagnostics will be output to `Stderr`. To see tracer startup logs, either add a logger, or set `DATADOG_TRACE_DEBUG=true` in your configuration and run your application with `dd-trace-run`.  This will automatically add a logger, and will expose both debug and startup tracer logs.

Reviewer: ```suggestion
If there is no logging configuration, only Diagnostics will be output to `Stderr`. To see tracer startup logs, either add a logger, or set `DATADOG_TRACE_DEBUG=true` in your configuration and run your application with `dd-trace-run`.  This adds a logger, and exposes both debug and startup tracer logs.
```

```

#### Diagnostics

> The Python Tracer has one diagnostic line it may print, in the case the agent cannot be reached.

Reviewer: ```suggestion
The Python tracer prints a diagnostic line when the Agent cannot be reached.
```

```text
DATADOG TRACER DIAGNOSTIC - Agent not reachable. Exception raised: [Errno 61] Connection refused
```

{{% /tab %}}
{{% tab "Ruby" %}}

#### Configuration

```text
W, [2020-07-08T21:14:25.281615 #137]  WARN -- ddtrace: [ddtrace] DATADOG TRACER CONFIGURATION - {"date":"2020-07-08T21:14:25+00:00","os_name":"x86_64-pc-linux-gnu","version":"0.37.0","lang":"ruby","lang_version":"2.7.0","enabled":true,"agent_url":"http://ddagent:8126?timeout=1","debug":false,"analytics_enabled":false,"runtime_metrics_enabled":false,"vm":"ruby-2.7.0","partial_flushing_enabled":false,"priority_sampling_enabled":false,"health_metrics_enabled":false}
```

#### Diagnostics

> The Ruby Tracer has one diagnostic line it may print, in the case the agent cannot be reached.

Reviewer: ```suggestion
The Ruby tracer prints a diagnostic line when the Agent cannot be reached.
```

```text
W, [2020-07-08T21:19:05.765994 #143]  WARN -- ddtrace: [ddtrace] DATADOG TRACER DIAGNOSTIC - Agent Error: Datadog::Transport::InternalErrorResponse ok?: unsupported?:, not_found?:, client_error?:, server_error?:, internal_error?:true, payload:, error_type:Errno::ECONNREFUSED error:Failed to open TCP connection to ddagent:9127 (Connection refused - connect(2) for "ddagent" port 9127)
```

{{% /tab %}}
{{% tab "C++" %}}

#### File Location

> In C++, the startup log file is created under the location `/var/tmp/dd-opentracing-cpp` e.g. `/var/tmp/dd-opentracing-cpp/startup_options-1593737077369521386.json`

Reviewer: ```suggestion
For C++, the startup log file is created in `/var/tmp/dd-opentracing-cpp`, for example: `/var/tmp/dd-opentracing-cpp/startup_options-1593737077369521386.json`.
```

#### Configuration

```text

{"agent_url":"http://localhost:8126","analytics_enabled":false,"analytics_sample_rate":null,"date":"2020-07-03T00:44:37+0000","dd_version":"","enabled":true,"env":"test-env","lang":"cpp","lang_version":"201402","operation_name_override":"","report_hostname":false,"sampling_rules":"[{\"sample_rate\": 1.0}]","service":"service_name","tags":{},"version":"v1.2.0"}

```

#### Diagnostics

> For C++, there are no `DATADOG TRACER DIAGNOSTICS` lines output to the tracer logs. However, if the Agent is not reachable, there should be error logs available, or in Envoy, an increase in the metrics `tracing.datadog.reports_failed` and `tracing.datadog.reports_dropped`.

Reviewer: ```suggestion
For C++, there are no `DATADOG TRACER DIAGNOSTICS` lines output to the tracer logs. However, if the Agent is not reachable, error logs are available, or in Envoy there is an increase in the metrics `tracing.datadog.reports_failed` and `tracing.datadog.reports_dropped`.
```

{{% /tab %}}
{{< /tabs >}}

## Diagnostics Errors

> The most insightful `DIAGNOSTICS` error output will be that the tracer is unable to send traces to the Datadog Agent.

Reviewer: ```suggestion
Look for `DIAGNOSTICS` error output that says the tracer is unable to send traces to the Datadog Agent.
```

> In this case, please confirm your tracer has been set up correctly to receive traces for [ECS][1], [Kubernetes][2], [Docker][3] or [any other option][4], or [contact support][5] who will be happy to review your Tracer & Agent configuration.

Reviewer: ```suggestion
If you have these errors, check that your tracer is set up to receive traces for [ECS][1], [Kubernetes][2], [Docker][3] or [any other option][4], or [contact support][5] to review your Tracer & Agent configuration.
```

## Configuration Settings

> If your logs only contain `CONFIGURATION` lines, a useful troubleshooting step is to confirm that the settings output by the tracer match the settings from your deployment and configuration of the Datadog Tracer.  Additionally, if you are not seeing specific traces in Datadog, please review the [Compatibility Requirements][6] section of the documentation to confirm these integrations are supported.

Reviewer: ```suggestion
If your logs contain only `CONFIGURATION` lines, a useful troubleshooting step is to confirm that the settings output by the tracer match the settings from your deployment and configuration of the Datadog Tracer.  Additionally, if you are not seeing specific traces in Datadog, review the [Compatibility Requirements][6] section of the documentation to confirm these integrations are supported.
```

> If there is an integration you are using that is not supported, or you want a fresh pair of eyes on your configuration output to understand why traces are not appearing as expected in Datadog, [contact support][5] who can help you diagnose and create a Feature Request for a new integration if needed.

Reviewer: ```suggestion
If an integration you are using is not supported, or you want a fresh pair of eyes on your configuration output to understand why traces are not appearing as expected in Datadog, [contact support][5] who can help you diagnose and create a Feature Request for a new integration.
```

## Disabling Startup Logs

> Per language, startup logs can be disabled if required via the environment variable setting `DD_TRACE_STARTUP_LOGS=false`.  It is not recommended to do this unless the logs emitted are posing a problem, and if [debug][7] logs are later sent, please make sure to enable startup logs and send all relevant logs together to speed up your support case triage.

Reviewer: ```suggestion
For each language, you can disable startup logs by setting the environment variable `DD_TRACE_STARTUP_LOGS=false`, but do this only if the logs emitted are posing a problem. If later you are sending [debug][7] logs, remember to enable startup logs and send all relevant logs together to speed up your support case triage.
```
